man/man1/plumb.1
.TH PLUMB 1
.SH NAME
plumb \- send message to plumber
.SH SYNOPSIS
.B plumb
[
.B -p
.I plumbfile
] [
.B -a
.I attributes
] [
.B -s
.I source
] [
.B -d
.I destination
] [
.B -t
.I type
] [
.B -w
.I directory
]
.B -i
|
.I data...
.SH DESCRIPTION
The
.I plumb
command formats and sends a plumbing message whose data
is, by default, the concatenation of the argument strings separated by blanks.
The options are:
.TP
.B -p
write the message to
.I plumbfile
(default
.BR /mnt/plumb/send ).
.TP
.B -a
set the
.B attr
field of the message (default is empty).
.TP
.B -s
set the
.B src
field of the message (default is
.BR plumb ).
.TP
.B -d
set the
.B dst
field of the message (default is empty).
.TP
.B -t
set the
.B type
field of the message (default is
.BR text ).
.TP
.B -w
set the
.B wdir
field of the message (default is the current working directory of
.IR plumb ).
.TP
.B -i
take the data from standard input rather than the argument strings.
If an
.B action=
attribute is not otherwise specified,
.I plumb
will add an
.B action=showdata
attribute to the message.
.SH FILES
.TF $HOME/lib/plumbing
.TP
.B $HOME/lib/plumbing
default rules file
.SH SOURCE
.B \*9/src/cmd/plumb
.SH "SEE ALSO"
.MR plumb 3 ,
.MR plumber 4 ,
.MR plumb 7


CHANGES
# Note: the name listed next to a change sometimes
# submitted the patch but more often just reported
# the problem.

# For entries more recent than January 30, 2008,
# see the Git repository history at 
# https://github.com/9fans/plan9port/.

January 30, 2008
	acme: fix regexp match choice bug
	devdraw: add F11 for full-screen toggle
	rio: accomodate full-screen apps more gracefully
	sam: fix regexp match choice bug

January 28, 2008
	xd: fix printing of 4-byte values on 64-bit machines (Sam Hopkins)

January 12, 2008
	delatex: reset lexer state at each beginning of file
	ipso: do not flush factotum keys unless using factotum file
	libdiskfs: fix ext2 directory reading bug (Mike Mammarella)
	lp: move devices to devices.sample
	rio: make full-screen work properly; add showevent
	venti: fewer calls to gettimeofday
	vmount: need to say -t nfs now on Linux

January 10, 2008
	acme: make STACK bigger for OS X (David Swasey)
	dist,unix: use MKSHELL=$PLAN9/bin/rc (John DeGood)
	libdraw: fix spinning event loop (David Bulkow)
	libregexp: fix match choice bug

December 7, 2007
	acme, sam, libregexp: regexp fixes and reversions

December 4, 2007
	hoc: print exact floating point output
	sam: turn off DEBUG logging (oops)
	vbackup: fix uninitialized variable use in -i code

November 28, 2007
	acme: refresh window properly after Tab command (Fazlul Shahriar)

November 27, 2007
	.hgignore: ignore more
	libdiskfs: make ext2 use little-endian disk format on big-endian machines (Joshua Wood)
	acme: accept clicks between body and scroll bar (Axel Belinfante)
	tpic: clean up better
	upas/smtp: remove intermediate files
	vbackup: add -i flag to avoid score prefetching (David Swasey)
	venti: add debugging HTML form

November 25, 2007
	acme: avoid infinite loop on search in 8125-byte file (Roger Peppe)
	lib9: avoid DIOCGMEDIASIZE on FreeBSD 4.11 (Rodrigo Readi)
	mailfs: use root in imapcopylist (Richard Bilson)
	moveplan9.files: add tmac/tmac.an (Tony Lainson)

November 23, 2007
	9pfuse: ignore Linux O_CLOEXEC flag (Michael Kaminsky)

November 22, 2007
	9pfuse: fix array bounds error (Michael Teichgräber)
	vbackup: fix fsysopen error message (Joshua Wood)

November 7, 2007
	9fs: use -n to mount sources (Jeff Sickel),
	9pfuse: MacFUSE moved again! (Jeff Sickel)

November 6, 2007
	.cvsignore: add 9pm (for old trees)
	unvf: ignore case in Content-Transfer-Encoding: line.
	validateattachment: reject malformed zip files

November 5, 2007
	building: use ar s flag to avoid ranlib (OS X)
	diff: continue after i/o errors when diffing many files
	factotum: silence gcc warning
	getns: turn / into _ in $DISPLAY
	libdiskfs: OS X linker fixes
	libmach: OS X linker fixes
	libmp: no text jump labels (OS X)
	libthread: use our own swapcontext on OS X
	mailfs: handle NIL body, add -r flag (Richard Bilson)

November 4, 2007
	man: document hg interface, fix CVS host name

November 3, 2007
	libthread: more OS X 10.5 changes (Jeff Sickel)

October 31, 2007
	venti: use 9ar, not ar (fixes Darwin build)

October 30, 2007
	libthread: OS X 10.5 tweaks (Jeff Sickel, Bakul Shah)

October 29, 2007
	grap: fclose, error context bug fixes (Doug Gwyn)
	venti: fix sync deadlock, add /proc stub

October 12, 2007
	9p: add -n to disable authentication
	9pserve: add -n to reject authentication
	srv: add -n to reject authentication

October 9, 2007
	mount: ask for Unix test explicitly (Chris King)

September 24-25, 2007
	venti: many fixes, new index cache

September 16, 2007
	factotum: fix build error (Axel Belinfante)

September 15, 2007
	auxstats: use acpi for battery on Linux (Andrey Mirtchovski)

September 13, 2007
	factotum: implement 9pcr (of dubious utility)

August 28, 2007
	9pserve: fix infinite loop bug

August 23, 2007
	venti: add -s flag, bug fixes (Michael Kaminsky)

August 22, 2007
	9.rc: change to work with rc -e (Tony Lainson)
	9pserve: fix races causing assert(c->nmsg == 0) to fail.
	acme: fix warning memory leak (Michael Teichgräber)
	libfmt: add fmtlocale to Makefile (Faried Nawaz)
	libthread: simplify alt tracking, possibly fixing bug
	mailfs: fix handling of FLAGS responses (Robert Vollmert)
	unix ports: Linux x86-64 Makefiles (Ævar Arnfjörð Bjarmason)

July 25, 2007
	9pfuse: even more MacFUSE path changes

June 29, 2007
	dict: look in $dictpath for dictionaries

June 24, 2007
	mailfs: fix subject line in header file.

June 23, 2007
	mailfs: various IMAP bug fixes
	Mail: add Get command, refresh at startup,
		fix deleted messages bug
	nedmail: refresh at startup, on i and y.

June 20, 2007
	9pfuse: add -a aname to get at sourcesdump

June 19, 2007
	9pfuse: the MacFUSE file locations changed again (Rob Pike)

June 18, 2007
	upas/marshal: remove debugging print
	9pserve: ignore window resizes (fix really should be elsewhere)
	libdiskfs: more HFS fixes (David Swasey)

June 15, 2007
	acme(1): fix Snarf bug, change Indent ON to apply to all windows.

June 12, 2007
	vcat: always write last block, to make disk correct size
	vbackup: more OS X fixes, add -M flag
	vbackup(8): document -M

June 9, 2007
	libregexp, sam, acme: day 1 bug in sam regexp code

June 8, 2007
	libdiskfs: add HFS resource fork support (David Swasey)

June 5, 2007
	ramfs: be less chatty

June 3, 2007
	vbackup: fix dotdot (David Swasey)

May 29, 2007
	libdiskfs: more HFS fixes (David Swasey)
	
May 28, 2007
	libdiskfs: missing Bterm, make ext2 quieter,
		add disk partitions, HFS code (David Swasey)

May 21, 2007
	libregexp: fix pointer comparison warning (Ramon de Vera)
	test: fix
	
May 20, 2007
	touch: print error if wstat fails
	libc: use futimes on Linux to implement dirfwstat

May 19, 2007
	test: import from Plan 9

May 18, 2007
	libregexp: fix bug about leading ^ in regexp on multiline text (Roger Peppe)

May 15, 2007
	libc: new media size code for _p9dir on FreeBSD (Lou Kamenov)
	acme(4): document tab width field in ctl file

May 10, 2007
	libc: more NaN fixes (rsc)
		use device names in _p9dir for FreeBSD (Lou Kamenov)

src/cmd/faces/plumb.c
#include <u.h>
#include <libc.h>
#include <thread.h>
#include <draw.h>
#include <regexp.h>
#include <bio.h>
#include <9pclient.h>
#include <plumb.h>
#include "faces.h"

static CFid*	showfd;
static CFid*	seefd;

char		**maildirs;
int		nmaildirs;

void
initplumb(void)
{
	showfd = plumbopenfid("send", OWRITE);
	seefd = plumbopenfid("seemail", OREAD);
	if(showfd == nil || seefd == nil)
		sysfatal("plumbopen: %r");
}

void
addmaildir(char *dir)
{
	maildirs = erealloc(maildirs, (nmaildirs+1)*sizeof(char*));
	maildirs[nmaildirs++] = dir;
}

char*
attr(Face *f)
{
	static char buf[128];

	if(f->str[Sdigest]){
		snprint(buf, sizeof buf, "digest=%s", f->str[Sdigest]);
		return buf;
	}
	return nil;
}

void
showmail(Face *f)
{
	char buf[256];
	Plumbmsg pm;
	Plumbattr a;

	if(showfd<0 || f->str[Sshow]==nil || f->str[Sshow][0]=='\0')
		return;
	snprint(buf, sizeof buf, "Mail/%s", f->str[Sshow]);
	pm.src = "faces";
	pm.dst = "showmail";
	pm.wdir = "/";
	pm.type = "text";
	a.name = "digest";
	a.value = f->str[Sdigest];
	a.next = nil;
	pm.attr = &a;
	pm.ndata = strlen(buf);
	pm.data = buf;
	plumbsendtofid(showfd, &pm);
}

char*
value(Plumbattr *attr, char *key, char *def)
{
	char *v;

	v = plumblookup(attr, key);
	if(v)
		return v;
	return def;
}

void
setname(Face *f, char *sender)
{
	char *at, *bang;
	char *p;
	char *fld[3];
	int nf;

	p = estrdup(sender);
	nf = tokenize(p, fld, 3);
	if(nf <= 1)
		sender = estrdup(fld[0]);
	else
		sender = estrdup(fld[1]);
	free(p);

	/* works with UTF-8, although it's written as ASCII */
	for(p=sender; *p!='\0'; p++)
		*p = tolower(*p);
	f->str[Suser] = sender;
	at = strchr(sender, '@');
	if(at){
		*at++ = '\0';
		f->str[Sdomain] = estrdup(at);
		return;
	}
	bang = strchr(sender, '!');
	if(bang){
		*bang++ = '\0';
		f->str[Suser] = estrdup(bang);
		f->str[Sdomain] = sender;
		return;
	}
}

static char* months[] = {
	"jan", "feb", "mar", "apr",
	"may", "jun", "jul", "aug",
	"sep", "oct", "nov", "dec"
};

static int
getmon(char *s)
{
	int i;

	for(i=0; i<nelem(months); i++)
		if(cistrcmp(months[i], s) == 0)
			return i;
	return -1;
}

/* Fri Jul 23 14:05:14 EDT 1999 */
ulong
parsedatev(char **a)
{
	char *p;
	Tm tm;

	memset(&tm, 0, sizeof tm);
	if((tm.mon=getmon(a[1])) == -1)
		goto Err;
	tm.mday = strtol(a[2], &p, 10);
	if(*p != '\0')
		goto Err;
	tm.hour = strtol(a[3], &p, 10);
	if(*p != ':')
		goto Err;
	tm.min = strtol(p+1, &p, 10);
	if(*p != ':')
		goto Err;
	tm.sec = strtol(p+1, &p, 10);
	if(*p != '\0')
		goto Err;
	if(strlen(a[4]) != 3)
		goto Err;
	strcpy(tm.zone, a[4]);
	if(strlen(a[5]) != 4)
		goto Err;
	tm.year = strtol(a[5], &p, 10);
	if(*p != '\0')
		goto Err;
	tm.year -= 1900;
	return tm2sec(&tm);
Err:
	return time(0);
}

ulong
parsedate(char *s)
{
	char *f[10];
	int nf;

	nf = getfields(s, f, nelem(f), 1, " ");
	if(nf < 6)
		return time(0);
	return parsedatev(f);
}


char*
tweakdate(char *d)
{
	char e[8];

	/* d, date = "Mon Aug  2 23:46:55 EDT 1999" */

	if(strlen(d) < strlen("Mon Aug  2 23:46:55 EDT 1999"))
		return estrdup("");
	if(strncmp(date, d, 4+4+3) == 0)
		snprint(e, sizeof e, "%.5s", d+4+4+3);	/* 23:46 */
	else
		snprint(e, sizeof e, "%.6s", d+4);	/* Aug  2 */
	return estrdup(e);
}

Face*
nextface(void)
{
	int i;
	Face *f;
	Plumbmsg *m;
	char *t, *data, *showmailp, *digestp;
	ulong xtime;

	f = emalloc(sizeof(Face));
	for(;;){
		m = plumbrecvfid(seefd);
		if(m == nil)
			killall("error on seemail plumb port");
		if(strncmp(m->data, "Mail/", 5) != 0){
			plumbfree(m);
			continue;
		}
		t = value(m->attr, "srvname", srvname);
		if(strcmp(t, srvname) != 0){
			plumbfree(m);
			continue;
		}
		data = m->data+5;
		t = value(m->attr, "mailtype", "");
		if(strcmp(t, "delete") == 0)
			delete(data, value(m->attr, "digest", nil));
		else if(strcmp(t, "new") != 0)
			fprint(2, "faces: unknown plumb message type %s\n", t);
		else for(i=0; i<nmaildirs; i++)
			if(strncmp(data, maildirs[i], strlen(maildirs[i])) == 0)
				goto Found;
		plumbfree(m);
		continue;

	Found:
		xtime = parsedate(value(m->attr, "date", date));
		digestp = value(m->attr, "digest", nil);
		if(alreadyseen(digestp)){
			/* duplicate upas/fs can send duplicate messages */
			plumbfree(m);
			continue;
		}
		showmailp = estrdup(data);
		if(digestp)
			digestp = estrdup(digestp);
		setname(f, value(m->attr, "sender", "???"));
		plumbfree(m);
		f->time = xtime;
		f->tm = *localtime(xtime);
		f->str[Sshow] = showmailp;
		f->str[Sdigest] = digestp;
		return f;
	}
	return nil;
}

char*
iline(char *data, char **pp)
{
	char *p;

	if(*data == 0)
		return nil;
	for(p=data; *p!='\0' && *p!='\n'; p++)
		;
	if(*p == '\n')
		*p++ = '\0';
	*pp = p;
	return data;
}

Face*
dirface(char *dir, char *num)
{
	Face *f;
	char buf[1024],  *info, *p, *t, *s;
	int n;
	ulong len;
	CFid *fid;

	sprint(buf, "%s/%s/info", dir, num);
	len = fsdirlen(mailfs, buf);
	if(len <= 0)
		return nil;
	fid = fsopen(mailfs, buf, OREAD);
	if(fid == nil)
		return nil;
	info = emalloc(len+1);
	n = fsreadn(fid, info, len);
	fsclose(fid);
	if(n < 0){
		free(info);
		return nil;
	}
	info[n] = '\0';
	f = emalloc(sizeof(Face));
	for(p=info; (s=iline(p, &p)) != nil; ){
		t = strchr(s, ' ');
		if(t == nil)
			continue;
		*t++ = 0;
		if(strcmp(s, "unixdate") == 0){
			f->time = atoi(t);
			f->tm = *localtime(f->time);
		}
		else if(strcmp(s, "from") == 0)
			setname(f, t);
		else if(strcmp(s, "digest") == 0)
			f->str[Sdigest] = estrdup(t);
	}
	sprint(buf, "%s/%s", dir, num);
	f->str[Sshow] = estrdup(buf);
	free(info);
	return f;
}
